src/System.Management.Automation/engine/runtime/Binding/Binders.cs

```diff
@@ -6513,7 +6513,7 @@ public override DynamicMetaObject FallbackInvokeMember(DynamicMetaObject target,
             }
 
             var psMethod = methodInfo as PSMethod;
-            if (psMethod != null)
+            if (psMethod != null && !IsListForeachMagicMethod(methodInfo, target, args))
```

Comment:
Looking through the [reference source](https://referencesource.microsoft.com/#q=ForEach) - maybe this is the only case that matters, but there could be other classes that this affects, so we couldn't consider this a complete fix.

Some downsides to this approach:
* Only works for `System.Collections.Generic.List[T]`
* Doesn't work for any (unlikely) new overloads

Personally I'd rather see the more general approach we discussed on Gitter - to add proper extension method support so we can add `ForEach` and `Where` as extension methods and then rely on overload resolution to do the right thing depending on the arguments.

I do see we'd need to add some overloads for `ForEach` - the `expression` argument is currently `object`, but we'd want 3 overloads, `ScriptBlock`, `Type`, and `string`.